File: practices/chapter13_graph/python/ch_13_banana_dealer.py
import re

def check_is_banana_dealer(name=''):
  if re.match('banana', name, re.I):
    return True
  return False

# 以某人為起點，查他的朋友是否有人是香蕉經銷商
def search_banana_dealer(name, graph):
  searched_list = []

  dealers = []
  dealers += graph[name]
  banana_dealer_found_res = False
  banana_dealer_name = ''

  while dealers:
    person = dealers.pop(0)
    if check_is_banana_dealer(person):
      banana_dealer_found_res = True
      banana_dealer_name = person
      break
    else:
      searched_list.append(person)
      dealers += graph[person]
  return {
    'searched_list': searched_list,
    'banana_dealer_found_res': banana_dealer_found_res,
    'banana_dealer_name': banana_dealer_name
  }

people_graph = {
  'Tom': ['Ivan', 'Ira', 'Kevin'],
  'Ivan': ['Peter'],
  'Ira': ['Banana_Lin'],
  'Kevin': ['Mary'],
  'Peter': [],
  'Banana_Lin': [],
  'Mary': [],
}

def main():
  searched_res = search_banana_dealer('Tom', people_graph)
  print('Searched list: ', searched_res)
  # print('Banana dealer name: ', searched_res.banana_dealer_name)

if __name__ == "__main__":
    main()

File: practices/chapter13_graph/js/ch_13_banana_dealer.js
class BananaDealerSearch {
  constructor(graph) {
    this.graph = graph
  }

  checkIsBananaDealer(name='') {
    console.log(name)
    if(name.match(/banana/gi)) {
      return true
    }
    return false
  }

  searchDealer(startPerson='') {
    let searchedList = []
    let dealers = []
    dealers.push(...this.graph[startPerson])

    const res = {
      banana_dealer_found_res: false,
      banana_dealer_name: '',
    }

    while(dealers.length > 0) {
      const person = dealers.shift()
      if(this.checkIsBananaDealer(person)) {
        res.banana_dealer_found_res = true
        res.banana_dealer_name = person
        break
      } else {
        searchedList.push(person)
        dealers.push(...this.graph[person])
      }
    }

    return res
  }
}

const friendsGraph = {
  'Tom': ['Ivan', 'Ira', 'Kevin'],
  'Ivan': ['Peter'],
  'Ira': ['Banana_Lin'],
  'Kevin': ['Mary'],
  'Peter': [],
  'Banana_Lin': [],
  'Mary': [],
}

function main() {
  const res = new BananaDealerSearch(friendsGraph).searchDealer('Tom')
  console.log('Searched Result: ', res)
}

main()